doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,90 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _codegen_intro:
+
+Introduction
+############
+
+Python snippets that are inlined in a source file are used as code generators.
+The tool to scan the source file for the Python snippets and process them is
+Codegen. Codegen and part of this documentation is based on
+`Cog <https://nedbatchelder.com/code/cog/index.html>`_ from Ned Batchelder.
+
+The processing of source files is controlled by the CMake extension functions:
+zephyr_sources_codegen(..) or zephyr_library_sources_codegen(..). The generated
+source files are added to the Zephyr sources. During build the source files are
+processed by Codegen and the generated source files are written to the CMake
+binary directory.
+
+The inlined Python snippets can contain any Python code, they are regular
+Python scripts. All Python snippets in a source file and all Python snippets of
+included template files are treated as a python script with a common set of
+global Python variables. Global data created in one snippet can be used in
+another snippet that is processed later on. This feature could be used, for
+example, to customize included template files.
+
+An inlined Python snippet can always access the codegen module. The codegen
+module encapsulates and provides all the functions to retrieve information
+(options, device tree properties, CMake variables, config properties) and to
+output the generated code.
+
+Codegen transforms files in a very simple way: it finds chunks of Python code
+embedded in them, executes the Python code, and places its output combined with
+the original file into the generated file. The original file can contain
+whatever text you like around the Python code. It will usually be source code.
+
+For example, if you run this file through Codegen:
+
+::
+
+    /* This is my C file. */
+    ...
+    /**
+     * @code{.codegen}
+     * fnames = ['DoSomething', 'DoAnotherThing', 'DoLastThing']
+     * for fn in fnames:
+     *     codegen.outl("void %s();" % fn)
+     * @endcode{.codegen}
+     */
+    /** @code{.codeins}@endcode */
+    ...
+
+it will come out like this:
+
+::
+
+    /* This is my C file. */
+    ...
+    /**
+     * @code{.codegen}
+     * fnames = ['DoSomething', 'DoAnotherThing', 'DoLastThing']
+     * for fn in fnames:
+     *     codegen.outl("void %s();" % fn)
+     * @endcode{.codegen}
+     */
+    void DoSomething();
+    void DoAnotherThing();
+    void DoLastThing();
+    /** @code{.codeins}@endcode */
+    ...
+
+Lines with ``@code{.codegen}`` or ``@code{.codeins}@endcode`` are marker lines.
+The lines between ``@code{.codegen}`` and ``@endcode{.codegen}`` are the
+generator Python code. The lines between ``@endcode{.codegen}`` and
```

Comment:
The output is filled between ``@endcode{.codegen}`` and ``@code{.codeins}@endcode{.codeins}``. There is no output currently.

Reply:
Maybe a silly question, but I mention this because I'm wondering how careful folks need to be about using the exact layout you give in your examples, and what happens if they add comments around these constructs.  For example, what will the output look like if I were to add some more comments after the @endcode{.codegen} or I futz around with the @code{.codeins}@endcode comment block? Are my comments preserved in the generated output?

```
/* This is my C file. */
    ...
    /**
     * @code{.codegen}
     * fnames = ['DoSomething', 'DoAnotherThing', 'DoLastThing']
     * for fn in fnames:
     *     codegen.outl("void %s();" % fn)
     * @endcode{.codegen}
     * Here are some more comments about the codegen block
     */
    /** 
     * And here is another comment
     * @code{.codeins}
     * @endcode 
     */
    ...
```

Reply:
> what will the output look like if I were to add some more comments after the @Endcode{.codegen} or I futz around with the @code{.codeins}@Endcode comment block? Are my comments preserved in the generated output?

Everything between ```@endcode{.codegen}``` and ```@code{.codeins}@endcode``` will be replaced by the generated code in the generated file. There will be no changes in the source file itself - it is preserved there.

You should not add comments after ```@endcode{.codegen}``` and/ or before  ```@code{.codeins}@endcode```. You should also not split the ```@code{.codeins}@endcode``` marker into two lines.

I did not test all the possible fuzz around. If some standard fuzzing patterns will emerge the parser  will have to be made more robust. Currently it assumes some well defined marker structure.